text/0000-relax-const-restrictions.md

```diff
@@ -0,0 +1,74 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC introduces a new concept called the "constness-context" of an execution of a function. A `const fn` is executed in a constant evaluation constness-context if it was called in a context where only `const fn` are allowed (like `static` initializers) or inside another `const fn` that was executed in a const constness-context. All other calls to a `const fn` (for example in `main`) are in a runtime constness-context.
+
+A `const fn` is now allowed to exhibit different behavior depending on it being called in a constant evaluation or runtime constness-context. Language features and standard library functions may also differ in behavior depending on the constness-context they have been used or called in.
+
+The constness-context of a function is therefore observable behavior.
+
+If a `const fn` is called in a runtime constness-context, no additional restrictions are applied to it, and it may do anything a non-`const fn` can (for example, calling into FFI).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are two main drawbacks to this RFC.
+
+Firstly, pure `const fn` under the old rules can be seen as a simple optimization opportunity for naive optimizers, as they could just reuse constant evaluation for `const fn` if the argument is known at compile time, even if the function is in a runtime context. This RFC makes such an optimization impossible. This is not seen as a problem by the author, as a more advanced optimizer (like LLVM) is able to remove these calls at compile time through means other than Rust's constant evaluation (inlining and constant folding). Also, a constant evaluation system can still evaluate executions in a runtime constness-context, as long as it behaves exactly like runtime.
+
+Secondly, with the current rules around `const fn` purity, unsafe code could choose to only sometimes call unknown `const fn` code at runtime depending on whether it needs the result. The author does not see this as a significant drawback, as this functionality is better served by language features that target this use case directly (like a `pure` attribute) and is therefore out of scope for the language feature of "functions evaluatable at compile time".
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+An alternative is to keep the current rules. This is bad because the current rules are highly restrictive and don't have a significant benefit to them.
+
+It would also be possible to allow them to behave differently, but keep the restrictions around purity and determinism at runtime. This would still allow unsafe code to treat `const fn` specially, but this is not seen as a desirable feature of `const fn`.
+
+# Prior art
+[prior-art]: #prior-art
```

Comment:
another prior art is that Rust *removed* pure functions and the discussion was always "const is not pure through the backdoor" as long as I can remember

Reply:
>  the discussion was always "const is not pure through the backdoor" as long as I can remember

To be fair, I don't think that is accurate. In fact former lang team member Centril definitely hoped for `const fn` to also mean 'pure'. I think this was a case of keeping our options open to see which alternative would bring the bigger benefits.

But I agree that the benefits of const_eval_select and float-in-const outweigh the benefit of having pure functions in Rust -- way more people came asking for being able to do more things in `const` than for being able to write unsafe code with a safe API that imposes purity restrictions on users.

Reply:
The [RFC for unreserving `pure`](https://rust-lang.github.io/rfcs/2421-unreservations-2018.html#1-effects) explicitly says
> Here, pure denotes a deterministic function -- but we already have const for more or less the same, and it is unlikely that we would introduce an effect (or restriction thereof) that is essentially const fn but not entirely. So this use case is unlikely to happen.

So `const` was considered to be roughly equivalent to `pure`. TheRFC was also written by Centril

Reply:
I take this to mean that Centril didn't want to have `pure` *separate* from `const fn`, but preferred to handle both of these concerns with one keyword/mechanism.